### DISCUSSION TITLE
STAC DataPipes

### DISCUSSION THREAD

**Discussion by weiji14** [weiji14/zen3geo#48] [Ideas]
To enable cloud-native, streaming machine learning data pipelines based on SpatioTemporal Asset Catalogs ([STAC](https://stacspec.org))!

A torch [DataPipe](https://github.com/pytorch/data/tree/v0.4.0#what-are-datapipes) is a way of doing [composition over inheritance](https://realpython.com/inheritance-composition-python/#choosing-between-inheritance-and-composition-in-python_1). The philosophy is to have each `DataPipe` [do one thing and do it well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well) similar to the UNIX philosophy of [pipe](https://en.wikipedia.org/wiki/Pipeline_(Unix))-ing one piece of text to another command. The pipe syntax also has parallels with the method chaining way of `pandas` (see [`pandas.DataFrame.pipe`](https://pandas.pydata.org/pandas-docs/version/1.5/reference/api/pandas.DataFrame.pipe.html)).

## :book: STAC Readers

There are 4 parts as per https://stacspec.org/en/about/stac-spec, and one idea to have individual DataPipes for each of the STAC Item/Catalog/Collection/API as hinted in https://github.com/microsoft/torchgeo/pull/412#issuecomment-1234584373

- [x] `PySTACItemReader` wrapping [`pystac.Item.from_file`](https://pystac.readthedocs.io/en/1.0/api/pystac.html#pystac.Item.from_file) (#46)
- [ ] `PySTACCatalogReader` wrapping [`pystac.Catalog.from_file`](https://pystac.readthedocs.io/en/1.0/api/pystac.html#pystac.Catalog.from_file) for static catalogs.
- [ ] `PySTACCollectionReader` wrapping [`pystac.ItemCollection.from_file`](https://pystac.readthedocs.io/en/1.0/api/pystac.html#pystac.ItemCollection.from_file)
- [x] `PySTACAPISearcher` wrapping e.g. [`pystac_client.Client.search`](https://pystac-client.readthedocs.io/en/v0.5.0/api.html#pystac_client.Client.search) for dynamic catalogs (#59)

See also https://stacspec.org/en/about/stac-spec/

## :floppy_disk: STAC I/O

Coming from the STAC Readers, the STAC objects (Item, ItemCollection, etc) would then need to be read into memory using some I/O library. These I/O libraries would handle the stacking of Assets as mentioned in https://github.com/microsoft/torchgeo/pull/412#issuecomment-1235731321. E.g.

- [x] `StackstacStacker` wrapping [`stackstac.stack`](https://stackstac.readthedocs.io/en/v0.4.2/api/main/stackstac.stack.html) which returns an `xarray.DataArray` (#61)
- [ ] `ODCstacLoader` wrapping [`odc.stac.load`](https://odc-stac.readthedocs.io/en/v0.2.2/_api/odc.stac.load.html) which returns an `xarray.Dataset`

*Note*: See also https://github.com/opendatacube/odc-stac/issues/54#issuecomment-1103313511 for differences between `stackstac` and `odc-stac`

## :service_dog: STAC services (requiring authentication)

- `planetary_computer` has their STAC catalog at https://planetarycomputer.microsoft.com/api/stac/v1/, and there are some (but not all) Collections which require signing/authentication
- `radiant-mlhub` has their own STAC catalog/API library, as mentioned in https://github.com/microsoft/torchgeo/pull/412#issuecomment-1233466183

Note: The authentication/signing can be handled via the `parameters` and/or `modifier` parameters in [`pystac_client.Client.open`](https://pystac-client.readthedocs.io/en/v0.5.0/api.html#pystac_client.Client.open) (I think).

## :cup_with_straw: Example 'DataPipeLine'

```mermaid
graph LR
    subgraph STAC DataPipeLine 1

    A["IterableWrapper (list[url])"] --> B
    B["PySTACItemReader (list[pystac.Item])"] --> C
    C["StackstacStacker (xarray.DataArray)"]

    end
```

## :people_holding_hands: Open for contributions

Anyone is welcome to comment on the details (e.g. naming the DataPipes, what else is needed, etc), or open a Pull Request directly to implement a DataPipe (see https://zen3geo.readthedocs.io/en/latest/CONTRIBUTING.html#running-things-locally on getting started)!

One thing to note is that I've designed `zen3geo` explicitly so that dependencies are [optional by default](https://github.com/weiji14/zen3geo/blob/600975a994c892705bbd849c233533bad0b73fb6/pyproject.toml#L24-L28), so if someone doesn't use `odc-stac` for example, they shouldn't have to install it. Just bear this in mind when you're writing up the code.

Cc @jamesvrt, @rbavery, @KennSmithDS

_Originally discussed in https://github.com/microsoft/torchgeo/pull/412, xref https://github.com/microsoft/torchgeo/issues/576_

**Comment by weiji14** [comment]
Heads up that there's now a MVP (minimal viable pipeline) from STAC API queries to a `stackstac`-created `xarray.DataArray` :tada:! I'm drafting a STAC DataPipe walkthrough tutorial at #62 (rendered preview at https://zen3geo--62.org.readthedocs.build/en/62/stacking.html). This will cover a DataPipeLine like so:

```mermaid
graph LR
    subgraph STAC DataPipeLine

    A["IterableWrapper (list[dict])"] --> B
    B["PySTACAPISearcher (list[pystac_client.ItemSearch])"] --> C
    C["Mapper (list[pystac.ItemCollection])"] --> D
    D["StackstacStacker (list[xarray.DataArray])"]

    end
```

where the steps are:

1. STAC queries written in the form of a Python `dict`
2. The `dict` queries get sent to `pystac_client.Client.search` and returns a `pystac_client.ItemSearch` instance
3. The `pystac_client.ItemSearch` gets turned into a `pystac.ItemCollection`
4. The `pystac.ItemCollection` object is passed to `stackstac.stack` and returns an `xarray.DataArray`

Hoping to finish this by the end of the week :crossed_fingers:, and will cut a new v0.5.0 release soon after :grin:

**Reply by weiji14** [reply]
STAC DataPipes are out in v0.5.0! See https://github.com/weiji14/zen3geo/discussions/68 for changelog :tada:

**Reply by rbavery** [reply]
I'll be testing this out and comparing it to a workflow where I'm using stackstac and xbatcher directly. Thanks a bunch for this @weiji14 !

**Reply by weiji14** [reply]
Cool, let me know if there are any bugs. I'm gonna start to convert one of my projects to use this too :grinning:

**Comment by weiji14** [comment]
Note that [zen3geo v0.6.0](https://github.com/weiji14/zen3geo/releases/tag/v0.6.0) comes with an [XpySTACAssetReader](https://zen3geo.readthedocs.io/en/v0.6.0/api.html#zen3geo.datapipes.XpySTACAssetReader) DataPipe for reading STAC assets backed by COG/NetCDF/Zarr files, done in https://github.com/weiji14/zen3geo/pull/87. This is essentially a wrapper around `xarray.open_dataset(..., engine="stac")` (requires [xpystac](https://github.com/stac-utils/xpystac/tree/0.0.1) to be installed), and enables reading a [`pystac.Asset`](https://pystac.readthedocs.io/en/1.0/api/pystac.html#pystac.Asset) directly into an [`xarray.Dataset`](https://docs.xarray.dev/en/v2023.04.2/generated/xarray.Dataset.html) object!
